New Data Source: aws_elb

bflad opened the pull request:
Similar to `aws_lb` resource and data source functionality, split `aws_elb` resource data management into its own "flatten" method and utilize it in this new data source.

Closes #1864

bflad commented:
This is ready for review, sorry for the extra commits.

jmehnle commented:
I'd appreciate if this got merged soon. (Others who agree, please don't post +1, but add a +1 reaction by clicking at the top right of my post.)

bflad commented:
Can I please get an initial review of this PR? Thanks!

radeksimko commented on aws/data_source_aws_elb_test.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+resource "aws_security_group" "elb_test" {
+  name        = "allow_all_elb_test"
```
Can you please randomize the name here, so it's safe to run this test multiple times in parallel and leaks don't cause subsequent runs to fail?

radeksimko commented on aws/data_source_aws_elb_test.go:
```diff
@@ -0,0 +1,113 @@
+  subnets         = ["${aws_subnet.elb_test.*.id}"]
+
+  idle_timeout = 30
+  enable_deletion_protection = false
```
This is an ALB/NLB specific field which needs to be removed. It's also the reason why the test is currently failing:

```
=== RUN   TestAccDataSourceAWSELB_basic
--- FAIL: TestAccDataSourceAWSELB_basic (3.64s)
	testing.go:503: Step 0 error: Configuration is invalid.

		Warnings: []string(nil)

		Errors: []string{"aws_elb.elb_test: : invalid or unknown key: enable_deletion_protection"}
```

radeksimko commented on aws/data_source_aws_elb_test.go:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  tags {
+    TestName = "TestAccAWSELB_basic"
```
Nitpick: it may be useful to make these tags in the test config data source specific.

bflad replied:
`t.Name()` to the rescue!

<img width="811" alt="screenshot 2017-12-06 19 31 22" src="https://user-images.githubusercontent.com/189114/33692617-1c6c854e-dabc-11e7-83a2-bdf3607cb063.png">

bflad commented on aws/resource_aws_elb.go:
```diff
@@ -348,24 +348,23 @@ func resourceAwsElbRead(d *schema.ResourceData, meta interface{}) error {
+
+// flattenAwsELbResource takes a *elbv2.LoadBalancer and populates all respective resource fields.
+func flattenAwsELbResource(d *schema.ResourceData, meta interface{}, lb *elb.LoadBalancerDescription) error {
+	elbconn := meta.(*AWSClient).elbconn
```
Unfortunately `sourceSGIdByName` requires `ec2conn`, but I will switch this to the two clients it needs.

bflad commented on aws/data_source_aws_elb.go:
```diff
@@ -0,0 +1,213 @@
+	log.Printf("[DEBUG] Reading ELBs: %#v", input)
+	resp, err := elbconn.DescribeLoadBalancers(input)
+	if err != nil {
+		return errwrap.Wrapf("Error retrieving LB: {{err}}", err)
```
Honestly, just copypasta. I'll fix it up. 👍

bflad commented:
Addressed comments and passes data source and resource testing for me! 🚀 
```
make testacc TEST=./aws TESTARGS='-run=\(TestAccDataSourceAWSELB\|TestAccAWSELB\)'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=\(TestAccDataSourceAWSELB\|TestAccAWSELB\) -timeout 120m
=== RUN   TestAccDataSourceAWSELB_basic
--- PASS: TestAccDataSourceAWSELB_basic (47.17s)
=== RUN   TestAccAWSELB_importBasic
--- PASS: TestAccAWSELB_importBasic (21.56s)
=== RUN   TestAccAWSELBAttachment_basic
--- PASS: TestAccAWSELBAttachment_basic (212.19s)
=== RUN   TestAccAWSELBAttachment_drift
--- PASS: TestAccAWSELBAttachment_drift (107.81s)
=== RUN   TestAccAWSELB_basic
--- PASS: TestAccAWSELB_basic (20.64s)
=== RUN   TestAccAWSELB_fullCharacterRange
--- PASS: TestAccAWSELB_fullCharacterRange (18.17s)
=== RUN   TestAccAWSELB_AccessLogs_enabled
--- PASS: TestAccAWSELB_AccessLogs_enabled (60.77s)
=== RUN   TestAccAWSELB_AccessLogs_disabled
--- PASS: TestAccAWSELB_AccessLogs_disabled (56.99s)
=== RUN   TestAccAWSELB_namePrefix
--- PASS: TestAccAWSELB_namePrefix (17.18s)
=== RUN   TestAccAWSELB_generatedName
--- PASS: TestAccAWSELB_generatedName (19.38s)
=== RUN   TestAccAWSELB_generatesNameForZeroValue
--- PASS: TestAccAWSELB_generatesNameForZeroValue (21.89s)
=== RUN   TestAccAWSELB_availabilityZones
--- PASS: TestAccAWSELB_availabilityZones (29.10s)
=== RUN   TestAccAWSELB_tags
--- PASS: TestAccAWSELB_tags (33.10s)
=== RUN   TestAccAWSELB_iam_server_cert
--- PASS: TestAccAWSELB_iam_server_cert (35.43s)
=== RUN   TestAccAWSELB_swap_subnets
--- PASS: TestAccAWSELB_swap_subnets (64.45s)
=== RUN   TestAccAWSELB_InstanceAttaching
--- PASS: TestAccAWSELB_InstanceAttaching (115.96s)
=== RUN   TestAccAWSELBUpdate_Listener
--- PASS: TestAccAWSELBUpdate_Listener (32.85s)
=== RUN   TestAccAWSELB_HealthCheck
--- PASS: TestAccAWSELB_HealthCheck (16.75s)
=== RUN   TestAccAWSELBUpdate_HealthCheck
--- PASS: TestAccAWSELBUpdate_HealthCheck (27.68s)
=== RUN   TestAccAWSELB_Timeout
--- PASS: TestAccAWSELB_Timeout (23.14s)
=== RUN   TestAccAWSELBUpdate_Timeout
--- PASS: TestAccAWSELBUpdate_Timeout (26.87s)
=== RUN   TestAccAWSELB_ConnectionDraining
--- PASS: TestAccAWSELB_ConnectionDraining (20.16s)
=== RUN   TestAccAWSELBUpdate_ConnectionDraining
--- PASS: TestAccAWSELBUpdate_ConnectionDraining (42.42s)
=== RUN   TestAccAWSELB_SecurityGroups
--- PASS: TestAccAWSELB_SecurityGroups (34.45s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1106.172s
```
